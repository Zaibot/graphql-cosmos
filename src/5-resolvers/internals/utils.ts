import { FeedResponse } from '@azure/cosmos'
import { GraphQLCosmosPageOutput } from '../../4-resolver-builder/3-typedefs-transformer'
import { Lazy } from '../../typescript'
import { SourceDescriptor } from '../x-descriptors'

export const emptyPageResponse: GraphQLCosmosPageOutput<any> = {
  cursor: null,
  nextCursor: null,
  page: [],
  total: 0,
}

export function graphqlCosmosPageResponse<T extends { id: string }>(
  cursor: string | null,
  query: Lazy<Promise<FeedResponse<T>>>,
  count: Lazy<Promise<FeedResponse<number>>>,
  transformer: (item: T) => SourceDescriptor.Embedded<T>
): GraphQLCosmosPageOutput<T> {
  return {
    cursor: cursor,
    nextCursor: () => query().then((x) => x.continuationToken),
    page: () => query().then((x) => (x.resources ?? []).map(transformer)),
    total: () => count().then((x) => x.resources[0]),
  }
}
