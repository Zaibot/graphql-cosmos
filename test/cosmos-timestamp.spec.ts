import gql from 'graphql-tag'
import { printSchemaWithDirectives } from 'graphql-tools'
import { createUnitTestContext } from './utils'

describe(`Processed schema`, () => {
  const dummyTypeDefs = gql`
    type Query {
      dummies: [Dummy] @cosmos
    }

    type Dummy @cosmos(database: "Test", container: "Dummies") {
      id: ID! @where(op: "eq in")
      timestamp: Int @cosmos(ours: "_ts") @where(op: "gte")
      etag: String @cosmos(ours: "_etag") @where(op: "eq")
    }
  `

  const responses = {
    Dummies: {
      'SELECT VALUE COUNT(1) FROM c': [1],
      'SELECT VALUE COUNT(1) FROM c WHERE c._etag = @p2 (@p2=abc)': [1],
      'SELECT VALUE COUNT(1) FROM c WHERE c._ts >= @p2 (@p2=100)': [1],

      'SELECT c.id FROM c ORDER BY c.id': [{ id: `1` }],
      'SELECT c.id FROM c WHERE c._etag = @p2 ORDER BY c.id (@p2=abc)': [{ id: `1` }],
      'SELECT c.id, c._ts, c._etag FROM c WHERE c._ts >= @p2 ORDER BY c.id (@p2=100)': [
        { id: `1`, _ts: 123, _etag: `abc` },
      ],

      'SELECT c.id, c._ts, c._etag FROM c WHERE ARRAY_CONTAINS(@p2, c.id) ORDER BY c.id (@p2=1)': [
        { id: `1`, _ts: 123, _etag: `abc` },
      ],
    },
  }

  const uc = createUnitTestContext(dummyTypeDefs, responses)

  it(`expects schema to remain the same`, () => {
    const output = printSchemaWithDirectives(uc.schema)
    expect(output).toMatchSnapshot()
  })

  it(`expects meta schema to remain the same`, () => {
    const output = uc.metaSchema
    expect(output).toMatchSnapshot()
  })

  it(`should return _ts _etag properties`, async () => {
    const result = await uc.execute(`query { dummies { total page { id timestamp etag } } } `)

    expect(result).toEqual({
      data: {
        dummies: { total: 1, page: [{ id: `1`, timestamp: 123, etag: `abc` }] },
      },
    })
  })

  it(`should be retrieve all items`, async () => {
    const result = await uc.execute(
      `query { dummies(where: { etag_eq: "abc" }) { total page { id timestamp etag } } } `
    )

    expect(result).toEqual({
      data: {
        dummies: { total: 1, page: [{ id: `1`, timestamp: 123, etag: `abc` }] },
      },
    })
  })

  it(`should be retrieve all items`, async () => {
    const result = await uc.execute(
      `query { dummies(where: { timestamp_gte: 100 }) { total page { id timestamp etag } } } `
    )

    expect(result).toEqual({
      data: {
        dummies: { total: 1, page: [{ id: `1`, timestamp: 123, etag: `abc` }] },
      },
    })
  })
})
