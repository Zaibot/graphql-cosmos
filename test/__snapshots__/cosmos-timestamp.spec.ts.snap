// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Processed schema expects meta schema to remain the same 1`] = `
Object {
  "types": Array [
    Object {
      "container": null,
      "cosmos": false,
      "database": null,
      "external": false,
      "fields": Array [
        Object {
          "container": null,
          "cosmos": true,
          "database": null,
          "fieldname": "etag",
          "ours": "_etag",
          "pagination": false,
          "returnMany": false,
          "returnTypename": "String",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [
            "eq",
          ],
          "whereOurs": null,
        },
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "id",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "ID",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [
            "eq",
            "in",
          ],
          "whereOurs": null,
        },
        Object {
          "container": null,
          "cosmos": true,
          "database": null,
          "fieldname": "timestamp",
          "ours": "_ts",
          "pagination": false,
          "returnMany": false,
          "returnTypename": "Int",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [
            "gte",
          ],
          "whereOurs": null,
        },
      ],
      "filterable": true,
      "sortable": false,
      "typename": "Dummy",
    },
    Object {
      "container": null,
      "cosmos": false,
      "database": null,
      "external": false,
      "fields": Array [
        Object {
          "container": "Dummies",
          "cosmos": true,
          "database": "Test",
          "fieldname": "dummies",
          "ours": null,
          "pagination": true,
          "returnMany": true,
          "returnTypename": "Dummy",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "filterable": false,
      "sortable": false,
      "typename": "Query",
    },
  ],
}
`;

exports[`Processed schema expects schema to remain the same 1`] = `
"schema {
  query: Query
}
type Query {
  dummies(where: DummyWhere, cursor: String, limit: Int): DummyPage!
}
type Dummy {
  id: ID!
  timestamp: Int
  etag: String
}
type DummyPage {
  total: Int!
  cursor: String
  nextCursor: String
  page: [Dummy]!
}
input DummyWhere {
  and: [DummyWhere!]
  or: [DummyWhere!]
  etag_eq: String
  id_eq: ID
  id_in: [ID]
  timestamp_gte: Int
}"
`;
